### ISSUE TITLE
OAuthAccessToken hasn't kept up with auth.py

### ISSUE THREAD

**Issue by athornton** [jupyterhub/jupyterhub#1309]
Hi! Thanks for using JupyterHub.

If you are reporting an issue with JupyterHub:

- Please use the [GitHub issue](https://github.com/jupyterhub/jupyterhub/issues)
  search feature to check if your issue has been asked already. If it has,
  please add your comments to the existing issue.

- Where applicable, please fill out the details below to help us troubleshoot
  the issue that you are facing. Please be as thorough as you are able to
  provide details on the issue.

**How to reproduce the issue**

Attempt to spawn an OAuth2-authenticated Lab session (with KubeSpawner).  OAuth2 is master commit 0927a19a2a3ddcf58a6938aa4297cd2df3ba474f plus my new auth_state logic, but nowhere in there does it touch the ORM stuff.

**What you expected to happen**

A container to spawn.

**What actually happens**
```
  Traceback (most recent call last):
      File "/usr/lib64/python3.4/site-packages/tornado/web.py", line 1509, in _execute
        result = method(*self.path_args, **self.path_kwargs)
      File "/usr/lib/python3.4/site-packages/jupyterhub/utils.py", line 193, in decorated
        return method(self, *args, **kwargs)
      File "/usr/lib/python3.4/site-packages/jupyterhub/apihandlers/auth.py", line 27, in get
        elif orm_token.service:
    AttributeError: 'OAuthAccessToken' object has no attribute 'service'
```
**Share what version of JupyterHub you are using**

Master, commit 0b34e13dd48bc73c8696eb63cde11b318b8f05a5

Running `jupyter troubleshoot` from the command line, if possible, and posting
its output would also be helpful.

```
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
$PATH:
	/usr/bin
	/usr/local/sbin
	/usr/local/bin
	/usr/sbin
	/usr/bin
	/sbin
	/bin

sys.path:
	/usr/bin
	/usr/lib64/python34.zip
	/usr/lib64/python3.4
	/usr/lib64/python3.4/plat-linux
	/usr/lib64/python3.4/lib-dynload
	/usr/lib64/python3.4/site-packages
	/usr/lib/python3.4/site-packages

sys.executable:
	/usr/bin/python3.4

sys.version:
	3.4.5 (default, May 29 2017, 15:17:55)
	[GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]

platform.platform():
	Linux-4.4.52+-x86_64-with-centos-7.3.1611-Core

pip list:
	alembic (0.9.3)
	asn1crypto (0.22.0)
	backports-abc (0.5)
	bleach (2.0.0)
	certifi (2017.7.27.1)
	cffi (1.10.0)
	chardet (3.0.4)
	cryptography (2.0.2)
	decorator (4.1.2)
	entrypoints (0.2.3)
	escapism (0.0.1)
	html5lib (0.999999999)
	httplib2 (0.10.3)
	idna (2.5)
	ipaddress (1.0.18)
	ipykernel (4.6.1)
	ipython (6.1.0)
	ipython-genutils (0.2.0)
	jedi (0.10.2)
	Jinja2 (2.9.6)
	jsonschema (2.6.0)
	jupyter-client (5.1.0)
	jupyter-core (4.3.0)
	jupyterhub (0.8.0.dev0)
	jupyterhub-kubespawner (0.5.1)
	jupyterlab (0.26.4)
	jupyterlab-launcher (0.3.1)
	kubernetes (2.0.0)
	Mako (1.0.7)
	MarkupSafe (1.0)
	mistune (0.7.4)
	nbconvert (5.2.1)
	nbformat (4.3.0)
	notebook (5.0.0)
	oauth2client (4.1.2)
	oauthenticator (0.7.0.dev0)
	pamela (0.3.0)
	pandocfilters (1.4.1)
	pexpect (4.2.1)
	pickleshare (0.7.4)
	pip (9.0.1)
	prompt-toolkit (1.0.15)
	ptyprocess (0.5.2)
	pyasn1 (0.3.2)
	pyasn1-modules (0.0.11)
	pycparser (2.18)
	pycurl (7.43.0)
	Pygments (2.2.0)
	python-dateutil (2.6.1)
	python-editor (1.0.3)
	python-oauth2 (1.0.1)
	PyYAML (3.12)
	pyzmq (16.0.2)
	requests (2.18.2)
	rsa (3.4.2)
	setuptools (36.2.5)
	simplegeneric (0.8.1)
	six (1.10.0)
	SQLAlchemy (1.1.12)
	terminado (0.6)
	testpath (0.3.1)
	tornado (4.5.1)
	traitlets (4.3.2)
	typing (3.6.1)
	urllib3 (1.22)
	wcwidth (0.1.7)
	webencodings (0.5.1)
	websocket-client (0.44.0)
	wheel (0.29.0)
```

*** Workaround ***

Patch jupyterhub/apihandlers/auth.py with

```
diff -u a/auth.py b/auth.py
--- a/auth.py	2017-08-04 16:21:55.000000000 -0700
+++ b/auth.py	2017-08-04 16:25:36.000000000 -0700
@@ -24,7 +24,7 @@
             raise web.HTTPError(404)
         if orm_token.user:
             model = self.user_model(self.users[orm_token.user])
-        elif orm_token.service:
+        elif "service" in orm_token and orm_token.service:
             model = self.service_model(orm_token.service)
         self.write(json.dumps(model))
```

**Comment by minrk** [comment]
Thanks. I can Something else is wrong, because there should be no such thing as an OAuthAccessToken without a user, and orm_token.service should only be retrieved if orm_token.user is None. The patch given should also result in a NameError on `model` because if neither is True, model with be undefined in `json.dumps(model)`.

@athornton can you provide `--debug` logs from JupyterHub surrounding the event?

**Comment by athornton** [comment]
Doesn't seem to be happening anymore in `ac33ba6ff4ee42f265e40dabbceb718003a6622e` .

**Comment by minrk** [comment]
Thanks! I'll close here , but feel free to reopen if you see something like this again.

I *suspect*, that this was fixed by 31f63264b0a86ed40d2f5738f02ef5cd03bd20b6, which ensures that the OAuth provider and handlers are using the same Session.
